# TASK 4.8 _ Grouping Data & Aggregating Variables

### 1) Create a new notebook with ords_prods_merge (with newly created columns from previous exercise) 

### 2) Find the aggregated mean of the “order_number” column grouped by “department_id” for the entire dataframe.

### 3) Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

### 4) Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

### 5) The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

### 6) The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
##### If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
##### If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

### 7) In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
##### If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
##### If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
##### If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

### 9) Export the dataframe as a pickle file in "Prepared Data" folder.

In [1]:
# Importing libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Creating path

path = r'C:\Users\Güldeniz\Desktop\Career Foundry\2. Data Immersion\Achievement 4\Instacart  Basket Analysis'

### 1) Import orders_products_merge from previous exercise.

In [3]:
# Import orders_products_merged:

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [6]:
ords_prods_merge.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     0.0         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  \
0          0         Soda        77              7     9.0   both   
1          1         Soda        77              7     9.0   both   
2          1         Soda        77              7     9.0   both   
3          1         Soda        77              7     9.0   both   
4          1         Soda        77              7     9.0   both   

     price_range_loc     busiest_day busiest_and_slowest_days  \
0  Mid-range product  Regularly busy           Regularly busy   
1  Mid-range product  Regularly busy             Slowest days   
2  Mid-range product  Regularly busy             Slowest days   
3  Mid-range product      Least busy             Slowest days   
4  Mid-range product      Least busy             Slowest days   

  busiest_period_of_day  
0        Average orders  
1        Average orders  
2           Most orders  
3        Average orders  
4           Most orders

### 2) Find aggregated mean of "order_number" grouped by "department_id".

In [7]:
ords_prods_merge ["order_id"] = ords_prods_merge["order_id"].astype("int32")

In [8]:
ords_prods_merge ["user_id"] = ords_prods_merge["user_id"].astype("int8")

In [9]:
ords_prods_merge ["order_number"] = ords_prods_merge["order_number"].astype("int8")

In [10]:
ords_prods_merge ["orders_day_of_week"] = ords_prods_merge["orders_day_of_week"].astype("int8")

In [11]:
ords_prods_merge ["order_hour_of_day"] = ords_prods_merge["order_hour_of_day"].astype("int8")

In [12]:
ords_prods_merge ["days_since_prior_order"] = ords_prods_merge["days_since_prior_order"].astype("int8")

In [13]:
ords_prods_merge ["product_id"] = ords_prods_merge["product_id"].astype("int8")

In [14]:
ords_prods_merge ["aisle_id"] = ords_prods_merge["aisle_id"].astype("int8")

In [15]:
ords_prods_merge ["department_id"] = ords_prods_merge["department_id"].astype("int8")

In [28]:
ords_prods_merge ["prices"] = ords_prods_merge["prices"].astype("int32")

In [17]:
ords_prods_merge ["product_name"] = ords_prods_merge["product_name"].astype("category")

In [18]:
ords_prods_merge ["price_range_loc"] = ords_prods_merge["price_range_loc"].astype("category")

In [19]:
ords_prods_merge ["busiest_and_slowest_days"] = ords_prods_merge["busiest_and_slowest_days"].astype("category")

In [20]:
ords_prods_merge ["busiest_period_of_day"] = ords_prods_merge["busiest_period_of_day"].astype("category")

In [21]:
#Single aggregation:

##Split the data into groups based on “department_id.”
##Apply the agg() function to each group to obtain the mean values for the “order_number” column.

ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

### 3) Analyze Results

##### With the entire dataframe, we include not only a selected number of departments, but all departments by their id numbers. Also when we compare the order number averages, we can easily observe that the mean values differ. For example, the average orders in department 4 equals to 18.8 in a subset with 1000000 observations, while the average value is 17.8 when the whole dataframe is checked.

### 4) Create a loyalty flag for existing customers using the transform() and loc() functions.

#### Aggregating data

In [22]:
# Split the data into groups based on the “user_id” column.=> df_ords_prods_merged.groupby(['user_id'])
## Apply the transform() function on the “order_number” column to generate the maximum orders for each user.=> ['order_number'].transform(np.max)
### Create a new column, “max_order,” into which you’ll place the results of your aggregation.=> df_ords_prods_merged['max_order'

ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [8]:
ords_prods_merge.head(15)

user_id  order_number  order_hour_of_day  days_since_prior_order  \
0         1             1                  8                     0.0   
1         1             2                  7                    15.0   
2         1             3                 12                    21.0   
3         1             4                  7                    29.0   
4         1             5                 15                    28.0   
5         1             6                  7                    19.0   
6         1             7                  9                    20.0   
7         1             8                 14                    14.0   
8         1             9                 16                     0.0   
9         1            10                  8                    30.0   
10       15            15                  9                     7.0   
11       15            17                 16                     8.0   
12       15            18                 11                     7.0   
13       15            21                  9                     7.0   
14       15            22                 10                    14.0   

    product_id product_name  department_id  prices    price_range_loc  \
0          196         Soda              7     9.0  Mid-range product   
1          196         Soda              7     9.0  Mid-range product   
2          196         Soda              7     9.0  Mid-range product   
3          196         Soda              7     9.0  Mid-range product   
4          196         Soda              7     9.0  Mid-range product   
5          196         Soda              7     9.0  Mid-range product   
6          196         Soda              7     9.0  Mid-range product   
7          196         Soda              7     9.0  Mid-range product   
8          196         Soda              7     9.0  Mid-range product   
9          196         Soda              7     9.0  Mid-range product   
10         196         Soda              7     9.0  Mid-range product   
11         196         Soda              7     9.0  Mid-range product   
12         196         Soda              7     9.0  Mid-range product   
13         196         Soda              7     9.0  Mid-range product   
14         196         Soda              7     9.0  Mid-range product   

   busiest_and_slowest_days busiest_period_of_day  max_order  
0            Regularly busy        Average orders         10  
1              Slowest days        Average orders         10  
2              Slowest days           Most orders         10  
3              Slowest days        Average orders         10  
4              Slowest days           Most orders         10  
5            Regularly busy        Average orders         10  
6              Busiest days           Most orders         10  
7              Busiest days           Most orders         10  
8              Busiest days           Most orders         10  
9              Slowest days        Average orders         10  
10             Busiest days           Most orders         22  
11           Regularly busy           Most orders         22  
12           Regularly busy           Most orders         22  
13             Busiest days           Most orders         22  
14             Busiest days           Most orders         22

#### Deriving columns with loc()

In [23]:
# If the maximum orders the user has made is over 40, then the customer will be labeled a “Loyal customer.”

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [24]:
# If the maximum orders the user has made is over 10 but less than or equal to 40, then the customer will be labeled a “Regular customer.”

ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [25]:
# If the maximum orders the user has made is less than or equal to 10, then the customer will be labeled a “New customer.”

ords_prods_merge.loc[ords_prods_merge['max_order'] <=10, 'loyalty_flag'] = 'New customer'

In [26]:
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Loyal customer    32404859
Name: loyalty_flag, dtype: int64

In [13]:
# Checking the basics:

ords_prods_merge.head()

user_id  order_number  order_hour_of_day  days_since_prior_order  \
0        1             1                  8                     0.0   
1        1             2                  7                    15.0   
2        1             3                 12                    21.0   
3        1             4                  7                    29.0   
4        1             5                 15                    28.0   

   product_id product_name  department_id  prices    price_range_loc  \
0         196         Soda              7     9.0  Mid-range product   
1         196         Soda              7     9.0  Mid-range product   
2         196         Soda              7     9.0  Mid-range product   
3         196         Soda              7     9.0  Mid-range product   
4         196         Soda              7     9.0  Mid-range product   

  busiest_and_slowest_days busiest_period_of_day  max_order  loyalty_flag  
0           Regularly busy        Average orders         10  New customer  
1             Slowest days        Average orders         10  New customer  
2             Slowest days           Most orders         10  New customer  
3             Slowest days        Average orders         10  New customer  
4             Slowest days           Most orders         10  New customer

In [14]:
#Checking selected columns

ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(50)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3
40       67  Regular customer             4
41       67  Regular customer             5
42       67  Regular customer             6
43       67  Regular customer             8
44       67  Regular customer             9
45       67  Regular customer            10
46       67  Regular customer            11
47       67  Regular customer            12
48       67  Regular customer            13
49       67  Regular customer            14

### 5) Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [29]:
# Basic statistics for customer royalty flag (loyal customer, reguar customer, new customer) by prices.
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices        
                    mean min max
loyalty_flag                    
Loyal customer  7.352975 -97  52

##### => The average price of the products purchased by loyal customers are lower (10.4) than the price of products purchased by new and regular customers .

### 6) Create a spending flag for each user based on the average price across all their orders using the following criteria:
##### a) If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
##### b) If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [11]:
ords_prods_merge.groupby('user_id').agg({'prices':'mean'}).head()

prices
user_id          
1        6.367797
2        7.515897
3        8.197727
4        8.205556
5        9.189189

In [12]:
# Split the data into groups based on the “user_id” column.
## Apply the transform() function on the “prices” column to generate the average orders for each user.
### Create a new column, “avg_price,” into which you’ll place the results of your aggregation.

ords_prods_merge['avg_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [18]:
ords_prods_merge.head()

user_id  order_number  order_hour_of_day  days_since_prior_order  \
0        1             1                  8                     0.0   
1        1             2                  7                    15.0   
2        1             3                 12                    21.0   
3        1             4                  7                    29.0   
4        1             5                 15                    28.0   

   product_id product_name  department_id  prices    price_range_loc  \
0         196         Soda              7     9.0  Mid-range product   
1         196         Soda              7     9.0  Mid-range product   
2         196         Soda              7     9.0  Mid-range product   
3         196         Soda              7     9.0  Mid-range product   
4         196         Soda              7     9.0  Mid-range product   

  busiest_and_slowest_days busiest_period_of_day  max_order  loyalty_flag  \
0           Regularly busy        Average orders         10  New customer   
1             Slowest days        Average orders         10  New customer   
2             Slowest days           Most orders         10  New customer   
3             Slowest days        Average orders         10  New customer   
4             Slowest days           Most orders         10  New customer   

   avg_price  
0   6.367797  
1   6.367797  
2   6.367797  
3   6.367797  
4   6.367797

#### Deriving columns with loc()

In [13]:
# a) Creating 'Low spender' spending flag for customers whose average spending is lower than 10.

ords_prods_merge.loc[ords_prods_merge['avg_price'] < 10, 'spending_flag'] = 'Low spender'

In [14]:
# b) Creating 'High spender' spending flag for customers whose average spending is higher than or equal to 10.

ords_prods_merge.loc[ords_prods_merge['avg_price'] >= 10, 'spending_flag'] = 'High spender'

In [15]:
# Checking the results
ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low spender     31770614
High spender      634245
Name: spending_flag, dtype: int64

In [23]:
ords_prods_merge[['user_id', 'spending_flag', 'prices']].head(50)

user_id spending_flag  prices
0         1   Low spender     9.0
1         1   Low spender     9.0
2         1   Low spender     9.0
3         1   Low spender     9.0
4         1   Low spender     9.0
5         1   Low spender     9.0
6         1   Low spender     9.0
7         1   Low spender     9.0
8         1   Low spender     9.0
9         1   Low spender     9.0
10       15   Low spender     9.0
11       15   Low spender     9.0
12       15   Low spender     9.0
13       15   Low spender     9.0
14       15   Low spender     9.0
15       19   Low spender     9.0
16       19   Low spender     9.0
17       19   Low spender     9.0
18       21   Low spender     9.0
19       31   Low spender     9.0
20       31   Low spender     9.0
21       43   Low spender     9.0
22       43   Low spender     9.0
23       52   Low spender     9.0
24       52   Low spender     9.0
25       52   Low spender     9.0
26       52   Low spender     9.0
27       52   Low spender     9.0
28       52   Low spender     9.0
29       52   Low spender     9.0
30       52   Low spender     9.0
31       52   Low spender     9.0
32       52   Low spender     9.0
33       52   Low spender     9.0
34       52   Low spender     9.0
35       52   Low spender     9.0
36       52   Low spender     9.0
37       67   Low spender     9.0
38       67   Low spender     9.0
39       67   Low spender     9.0
40       67   Low spender     9.0
41       67   Low spender     9.0
42       67   Low spender     9.0
43       67   Low spender     9.0
44       67   Low spender     9.0
45       67   Low spender     9.0
46       67   Low spender     9.0
47       67   Low spender     9.0
48       67   Low spender     9.0
49       67   Low spender     9.0

### 7) Determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
##### a) If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
##### b) If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
##### c) If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [16]:
# Split the data into groups based on the “user_id” column.
## Apply the transform() function on the “days_since_prior_order” column to generate the median orders for each user.
### Create a new column, “order_frequency,” into which the results of your aggregation will be placed.

ords_prods_merge['order_regularity'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

#### Deriving columns with loc()

In [17]:
# a) Creating a frequency flag called 'Non-frequent customer' for customers whose median of "days_since_last_order" is higher than 20.

ords_prods_merge.loc[ords_prods_merge['order_regularity'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [18]:
# b) Creating a frequency flag called 'Regular customer' for customers whose median of "days_since_last_order" is higher than 10 or equal to 20.


ords_prods_merge.loc[(ords_prods_merge['order_regularity'] <= 20) & (ords_prods_merge['order_regularity'] > 10 ), 'frequency_flag'] = 'Regular customer'

In [19]:
# a) Creating a frequency flag called 'Frequent customer' for customers whose median of "days_since_last_order" is lower or equal to 10.

ords_prods_merge.loc[ords_prods_merge['order_regularity'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [32]:
ords_prods_merge.head()

user_id  order_number  order_hour_of_day  days_since_prior_order  \
0        1             1                  8                     0.0   
1        1             2                  7                    15.0   
2        1             3                 12                    21.0   
3        1             4                  7                    29.0   
4        1             5                 15                    28.0   

   product_id product_name  department_id  prices    price_range_loc  \
0         196         Soda              7     9.0  Mid-range product   
1         196         Soda              7     9.0  Mid-range product   
2         196         Soda              7     9.0  Mid-range product   
3         196         Soda              7     9.0  Mid-range product   
4         196         Soda              7     9.0  Mid-range product   

  busiest_and_slowest_days busiest_period_of_day  max_order  loyalty_flag  \
0           Regularly busy        Average orders         10  New customer   
1             Slowest days        Average orders         10  New customer   
2             Slowest days           Most orders         10  New customer   
3             Slowest days        Average orders         10  New customer   
4             Slowest days           Most orders         10  New customer   

   avg_price spending_flag  order_regularity    frequency_flag  
0   6.367797   Low spender              20.0  Regular customer  
1   6.367797   Low spender              20.0  Regular customer  
2   6.367797   Low spender              20.0  Regular customer  
3   6.367797   Low spender              20.0  Regular customer  
4   6.367797   Low spender              20.0  Regular customer

In [20]:
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

Frequent customer        22796659
Regular customer          6921472
Non-frequent customer     2686728
Name: frequency_flag, dtype: int64

In [ ]:
ords_prods_merge ["loyalty_flag"] = ords_prods_merge["loyalty_flag"].astype("category")

In [ ]:
ords_prods_merge ["spending_flag"] = ords_prods_merge["spending_flag"].astype("category")

In [ ]:
ords_prods_merge ["frequency_flag"] = ords_prods_merge["frequency_flag"].astype("category")

### 9)Export as pkl

In [22]:
ords_prods_merge.to_pickle(os.path.join(path,'02 Data', 'Prepared Data', 'orders_products_merge_4.8.pkl'))